File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.util.Scanner;
import java.util.Stack;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] nums = new int[n];
        for (int i = 0; i < n; i++) {
            nums[i] = sc.nextInt();
        }
        
        Stack<Integer> stack = new Stack<>();
        int num = 1;
        StringBuffer bf = new StringBuffer();
        boolean isNoPrinted = false;
        
        for (int i = 0; i < nums.length; i++) {
            int cur = nums[i];
            
            if (cur >= num) {
                while (cur >= num) {
                    stack.push(num++);
                    bf.append("+\n");
                }
                
                stack.pop();
                bf.append("-\n");
            } else {
                int pop = stack.pop();
                if (pop > cur) {
                    System.out.println("NO");
                    isNoPrinted = true;
                    break;
                } else {
                    bf.append("-\n");
                }
            }
        }
        
        if (!isNoPrinted) {
            System.out.println(bf.toString());
        }
    }
}

File: 백준/Silver/1874. 스택 수열/스택 수열.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.Stack

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()
    val nums = IntArray(n)
    for (i in 0 until n) {
        nums[i] = br.readLine().toInt()
    }
    
    val stack = Stack<Int>()
    var num = 1
    val sb = StringBuilder()
    var isNoPrinted = false
    
    for (i in 0 until n) {
        val cur = nums[i]
        
        if (cur >= num) {
            while (cur >= num) {
                stack.push(num++)
                sb.append("+\n")
            }
            
            stack.pop()
            sb.append("-\n")
        } else {
            val pop = stack.pop()
            if (pop > cur) {
                println("NO")
                isNoPrinted = true
                break
            } else {
                sb.append("-\n")
            }
        }
    }
    
    if (!isNoPrinted) {
        print(sb.toString())
    }
}